flink-ml-lib/src/main/java/org/apache/flink/ml/feature/standardscaler/StandardScaler.java

```diff
@@ -0,0 +1,290 @@
+import org.apache.flink.api.java.typeutils.TupleTypeInfo;
+import org.apache.flink.iteration.operator.OperatorStateUtils;
+import org.apache.flink.ml.api.Estimator;
+import org.apache.flink.ml.linalg.BLAS;
+import org.apache.flink.ml.linalg.DenseVector;
+import org.apache.flink.ml.linalg.Vector;
+import org.apache.flink.ml.linalg.Vectors;
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.util.ParamUtils;
+import org.apache.flink.ml.util.ReadWriteUtils;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An Estimator which implements the standard scaling algorithm.
+ *
+ * <p>Standardization is a common requirement for machine learning training because they may behave
+ * badly if the individual features of a input do not look like standard normally distributed data
+ * (e.g. Gaussian with 0 mean and unit variance).
+ *
+ * <p>This estimator standardizes the input features by removing the mean and scaling each dimension
+ * to unit variance.
+ */
+public class StandardScaler
+        implements Estimator<StandardScaler, StandardScalerModel>,
+                StandardScalerParams<StandardScaler> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public StandardScaler() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public StandardScalerModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<Tuple3<DenseVector, DenseVector, Long>> sumAndSquaredSumAndWeight =
+                tEnv.toDataStream(inputs[0])
+                        .transform(
+                                "computeMeta",
+                                new TupleTypeInfo<>(
+                                        TypeInformation.of(DenseVector.class),
+                                        TypeInformation.of(DenseVector.class),
+                                        BasicTypeInfo.LONG_TYPE_INFO),
+                                new ComputeMeta(getFeaturesCol()));
```

Comment:
How about using the following structure here?
```java
.keyBy(<random_key>)
.windowAll()
.aggregate()/.reduce()
```
The expected behavior is to accumulate computed results and output when all inputs are finished, which I think can be completed by the code above. The code above also preserves parallelism of your job. Besides, it would use higher abstraction-level Flink API so you need not care about some implementation details. Using a keyBy also allows you to use `ValueState` instead of `ListState` + `getUniqueElement`.

Same for the following part of code about `BuildModel`.

Reply:
Good point. My concern is that using`windowAll/window` may lead to a large memory consumption. As I understand, window operations need to cache the data in state. (Please correct me if I am wrong...

Reply:
I think window operation would only cache data if the following operator is `apply`. In cases when the following is `aggregate/reduce`, the window operation processes a stream record as soon as it is received, so it would only need to cache the accumulator / reduce result.

Reply:
Thanks for the detail explanation. When using `keyBy().window().aggregate()`, I came into another problem [1]. It seems that it is not solved yet. So let's use the current implementation for now.

[1] https://lists.apache.org/thread/7jm3sdh2h3y2gks3rp8kn9yb44b55dzv